tensorflow/python/estimator/canned/dnn_linear_combined.py

```diff
@@ -179,31 +179,34 @@ def _dnn_linear_combined_model_fn(
   # Combine logits and build full model.
   if dnn_logits is not None and linear_logits is not None:
     logits = dnn_logits + linear_logits
+    model = "combined"
   elif dnn_logits is not None:
     logits = dnn_logits
+    model = "dnn"
   else:
     logits = linear_logits
+    model = "linear"
 
   def _train_op_fn(loss):
     """Returns the op to optimize the loss."""
-    train_ops = []
     global_step = training_util.get_global_step()
-    if dnn_logits is not None:
-      train_ops.append(
-          dnn_optimizer.minimize(
-              loss,
-              var_list=ops.get_collection(
-                  ops.GraphKeys.TRAINABLE_VARIABLES,
-                  scope=dnn_parent_scope)))
-    if linear_logits is not None:
-      train_ops.append(
-          linear_optimizer.minimize(
-              loss,
-              var_list=ops.get_collection(
-                  ops.GraphKeys.TRAINABLE_VARIABLES,
-                  scope=linear_parent_scope)))
-
-    train_op = control_flow_ops.group(*train_ops)
+
+    if model == "combined":
+      all_vars = ops.trainable_variables()
+      linear_vars = ops.get_collection(
+        ops.GraphKeys.TRAINABLE_VARIABLES,
+        scope=linear_parent_scope)
+      # variables left are optimized by dnn
+      dnn_vars = list(set(all_vars) - set(linear_vars))
```

Comment:
Perhaps it will be performance bottlenecks, better proposals are welcome and appreciated.